src/solution.c

```diff
@@ -409,15 +410,34 @@ static void solution_thread(void *arg)
     const channel_measurement_t *p_meas[n_ready];
     navigation_measurement_t *p_nav_meas[n_ready];
     const ephemeris_t *p_e_meas[n_ready];
+
+    /* Create navigation measurements from the channel measurements*/
+
+    /* Create arrays of pointers for use in calc_navigation_measurement */
     for (u8 i=0; i<n_ready; i++) {
       p_meas[i] = &meas[i];
       p_nav_meas[i] = &nav_meas[i];
       p_e_meas[i] = ephemeris_get(meas[i].sid);
     }
 
     ephemeris_lock();
-    calc_navigation_measurement(n_ready, p_meas, p_nav_meas,
-                                (double)((u32)nav_tc)/SAMPLE_FREQ, p_e_meas);
+    gps_time_t nav_time = rx2gpstime(nav_tc);
+    if (time_quality == TIME_FINE) {
+      /* If we have timing then we can calculate the relationship between
+       * receiver time and GPS time and hence provide the pseudorange
+       * calculation with the local GPS time of reception. */
+      calc_navigation_measurement(n_ready, p_meas, p_nav_meas,
```

Comment:
Conditionally define the `nav_time` pointer instead of duplicating everything:
`p_nav_time = (time_quality == TIME_FINE) ? &nav_time : NULL;`